client/allocrunner/csi_hook.go

```diff
@@ -107,6 +107,11 @@ func (c *csiHook) Postrun() error {
 
 	for _, pair := range c.volumeRequests {
 
+		err := c.unmount(pair)
+		if err != nil {
+			mErr = multierror.Append(mErr, err)
+		}
```

Comment:
Apologies if this was covered yesterday, but for my education, if we fail to unmount the volume, can attempting to unpublish the volume succeed or will it always fail? I am curious how the fall-through when an error is received works.

Reply:
Good catch. There are two cases:

* We may be returning an error that's [`structs.ErrCSIClientRPCIgnorable`](https://github.com/hashicorp/nomad/blob/v1.2.4/client/pluginmanager/csimanager/volume.go#L355). For example, this CSI RPC is supposed to be idempotent so if it's been node-unpublished already we still want to make the `CSIVolume.Unpublish` RPC back to the server so that the server can do a controller-unpublish and release the claim.
* If it's _not_ ignorable (ex. the node plugin's alloc is gone), we should probably _not_ send the `CSIVolume.Unpublish` because there's nothing we can do from the server that will improve things! Instead we should be retrying until it works with exponential backoff, and logging that so that the operator can intervene manually. That will cause alloc shutdown to hang, which I think is exactly what we want here.

That being said, I think we should push the "ignorable" errors down into the `UnmountVolume` so that we only ever return non-ignorable errors. Then this code can just have a happy path and a path where errors need to be retried and block this method from returning.

Reply:
@jrasell I've addressed this in https://github.com/hashicorp/nomad/pull/11892/commits/04f5b7aa29ec2b563a50e66cae93c25057a2f783:

> The CSI `NodeUnpublishVolume`/`NodeUnstageVolume` RPCs can return
ignorable errors in the case where the volume has already been
unmounted from the node. Handle all other errors by retrying until we
get success so as to give operators the opportunity to reschedule a
failed node plugin (ex. in the case where they accidentally drained a
node without `-ignore-system`). Fan-out the work for each volume into
its own goroutine so that we can release a subset of volumes if only
one is stuck.